runners/samza/src/main/java/org/apache/beam/runners/samza/metrics/SamzaMetricsContainer.java

```diff
@@ -81,15 +125,96 @@ public MetricsContainerStepMap getContainers() {
+        scheduledFuture.cancel(true);
+      }
+
+      // Update metrics one last time, ensuring that we're still committing from a single thread.
+      // This guarantees that all remaining metrics are committed before shutting down the executor
+      // service.
+      commitPeriodicMetricsForAllSteps();
+
+      // Shutdown the executor service gracefully.
+      // Allow any currently executing tasks to finish, then terminate the service. If the shutdown
+      // process takes longer than 5 seconds, force a shutdown to ensure the service is stopped.
+      scheduler.shutdown();
+      try {
+        if (!scheduler.awaitTermination(5, TimeUnit.SECONDS)) {
+          LOG.warn("Forcing shutdown of executor service...");
+          scheduler.shutdownNow();
+        }
+      } catch (InterruptedException e) {
+        LOG.error("Interrupted during shutdown, forcing shutdown now", e);
+        scheduler.shutdownNow();
+        Thread.currentThread().interrupt();
+      }
+    }
+  }
+
+  /**
+   * Commits metrics updates for all scheduled steps, invoked periodically.
+   *
+   * <p>This method is called periodically by the scheduled executor service when `deferToExecutor`
+   * is set to true, ensuring that metrics for all active steps are updated at the defined interval.
+   * It is also invoked during shutdown to ensure that any remaining metrics are committed before
+   * terminating the executor.
+   *
+   * <p>Synchronization is applied to ensure thread-safe updates when multiple instances might share
+   * the same resources.
+   */
+  private void commitPeriodicMetricsForAllSteps() {
+    synchronized (SamzaMetricsContainer.class) {
+      activeStepNames.forEach(this::updateMetricsInternal);
+    }
+  }
 
-    assert metricsRegistry != null;
-    final List<String> stepNameList = Arrays.asList(stepName, GLOBAL_CONTAINER_STEP_NAME);
-    // Since global metrics do not belong to any step, we need to update it in every step.
-    final MetricResults metricResults =
+  /**
+   * Updates metrics for a given step.
+   *
+   * <p>This method contains the core logic for updating metrics, including counters, gauges, and
+   * distributions for the specified step. It can be called either immediately or by the scheduled
+   * executor service, based on the `deferToExecutor` configuration.
+   *
+   * <p>This method assumes that it is called in a thread-safe manner, as it does not implement
+   * synchronization internally.
+   *
+   * @param stepName the step name for which metrics are being updated
+   */
+  private void updateMetricsInternal(String stepName) {
+    List<String> stepNameList = Arrays.asList(stepName, GLOBAL_CONTAINER_STEP_NAME);
+    MetricResults metricResults =
         asAttemptedOnlyMetricResultsForSteps(metricsContainers, stepNameList);
-    final MetricQueryResults results = metricResults.allMetrics();
+    MetricQueryResults results = metricResults.allMetrics();
```

Comment:
why the `final` is removed here?

Reply:
I think it was removed during `./gradlew :runners:samza:spotlessApply`, but it shouldn't matter since I was the one that introduced it originally. I'm actually big proponent of making everything final where possible so I can introduce it back if I ever make anymore changes in this class.